Source/Game/Game/Editor/AssetBrowser.cpp

```diff
@@ -0,0 +1,490 @@
+                textureDesc.path = item.string();
+                auto textureID = _renderer->LoadTexture(textureDesc);
+                _images[i] = _renderer->GetImguiImageHandle(textureID);
+                _imagesSize[i] = ImVec2((float)_renderer->GetTextureWidth(textureID),
+                    (float)_renderer->GetTextureHeight(textureID));
+            }
+
+            _currentImage = _images[i];
+            _currentSize = _imagesSize[i];
+            return;
+        }
+
+        _currentImage = _defaultImageHandle;
+        _currentSize = _defaultDisplaySize;
+    }
+
+    void AssetBrowser::CalculateAspectRatio(ImVec2 &size, const ImVec2 ref)
+    {
+        bool inverse = true;
+        float max = size.y;
+        float min = size.x;
+
+        if (size.x >= size.y)
+        {
+            max = size.x;
+            min = size.y;
+            inverse = false;
+        }
+
+        float ratio = max / ref.x;
+
+        // if the texture is too small,
+        // the resize ratio is blocked at x2.5
+        // to not make it too blurry
+        if (ratio < 0.4f)
+            ratio = 0.4f;
+
+        max /= ratio;
+        min /= ratio;
+
+        if (inverse)
+        {
+            size.x = min;
+            size.y = max;
+        }
+        else
+        {
+            size.x = max;
+            size.y = min;
+        }
+    }
+
+    bool AssetBrowser::SearchString(const std::string &ref, const std::string &key, bool insensitive)
+    {
+        std::string str = ref;
+        std::string substr = key;
+
+        if (insensitive)
+        {
+            std::transform(
+                str.begin(), str.end(), str.begin(), [](unsigned char c)
+                { return std::tolower(c); });
+            std::transform(
+                substr.begin(), substr.end(), substr.begin(), [](unsigned char c)
+                { return std::tolower(c); });
+        }
+
+        return (str.find(substr) != std::string::npos);
+    }
+
+    bool AssetBrowser::IsTexture(const std::string &ref)
+    {
+        return (ref == ".dds");
+    }
+
+    std::string AssetBrowser::EllipsisText(const char *ref, float width, float fontWidth)
+    {
+        ImVec2 textSize = ImGui::CalcTextSize(ref);
+        if (textSize.x > width)
+        {
+            float ellipsisWidth = ImGui::CalcTextSize("...").x;
+            int maxCharacters = static_cast<int>((width - ellipsisWidth) / (fontWidth * 1.1f));
+            return (std::string(ref).substr(0, maxCharacters - 1) + "...");
+        }
+
+        return ref;
+    }
+
+    void AssetBrowser::ProcessFileSearch(const std::string &search)
+    {
+        if (_searchedString != search)
+        {
+            _images.clear();
+            _imagesSize.clear();
+
+            if (search.size() > _searchedString.size() && !_searchedFiles.empty())
+            {
+                const auto temp = _searchedFiles;
+                _searchedFiles.clear();
+                _searchedFiles.shrink_to_fit();
```

Comment:
I don't think this shrink_to_fit is a good idea here. Since you just cleared the vector it's going to deallocate the memory of the vector, and then immediately reserve the new size. Just do the clear and the reserve so if the previous size was bigger it won't reallocate.